### ISSUE TITLE
Armeria server performance analaysis

### ISSUE THREAD

**Issue by ikhoon** [http4s/http4s-armeria#4]
A benchmark and performance analysis would be good data to advocate Armeria server to http4s' users.

**Comment by rossabaker** [comment]
@hamnis set up a [server-example project](https://github.com/hamnis/http4s-server-example), which we've used to run some crude benchmarks between the various backends.  It doesn't show off much yet, but it gives us a baseline for pings.

**Comment by ikhoon** [comment]
1st Benchmark:
- Environment: MacBook Pro, 2.2 GHz 6-Core Intel Core i7
- Benchmark code: https://github.com/hamnis/http4s-server-example/commit/0cfeee3e4c4fd965f23f04f40cdf8a70981285af
- Result
    - http4s-armeria
	  ```
	  $ wrk -t12 -c400 -d30s http://127.0.0.1:8080/hello
	  Running 30s test @ http://127.0.0.1:8080/hello
	    12 threads and 400 connections
	    Thread Stats   Avg      Stdev     Max   +/- Stdev
	      Latency    10.41ms    1.68ms  34.55ms   84.22%
	      Req/Sec     1.94k     0.92k    9.83k    60.23%
	    694307 requests in 30.10s, 115.25MB read
	    Socket errors: connect 155, read 106, write 0, timeout 0
	    Non-2xx or 3xx responses: 694307
	  Requests/sec:  23063.19
	  Transfer/sec:      3.83MB
	  ```
  - Blaze
	```
	$wrk -t12 -c400 -d30s http://127.0.0.1:8080/hello
	Running 30s test @ http://127.0.0.1:8080/hello
	  12 threads and 400 connections
	  Thread Stats   Avg      Stdev     Max   +/- Stdev
	    Latency     2.94ms    2.19ms 107.06ms   96.86%
	    Req/Sec     6.99k     3.06k   14.33k    53.22%
	  2506706 requests in 30.03s, 351.81MB read
	  Socket errors: connect 155, read 277, write 0, timeout 0
	Requests/sec:  83481.35
	Transfer/sec:     11.72MB
	```
- Bottleneck
  - http4s-armeria consumes all CPU cycles, it hits 100% 
  - http4s-blaze hits: 50~60%
- Analysis
  - The performance(Requests/sec) of http4s-armeria was increased **5 times** when removing `toUnicastPublisher` operation which converts `Stream[F, HttpObject]` to `Publihser[HttpObject]`
- Conclusion
  - We cannot release the initial version of http4s-armeria until solving the performance problem on the high CPU utilization.
  - We might need an obtimized converter for Reactive Streams?
  - ???

**Comment by ngbinh** [comment]
Is it resolved? Congrats on the new release!

**Comment by ikhoon** [comment]
> Is it resolved?

I think there is a room we can optimize http4s-armeria performance. 💪

Loads of the benchmark were a simple "hello world" text message with HTTP/1.1.
Because of the simple load, the bottleneck point which mostly consumes CPU was a conversion from Reactive Streams to fs2 `Stream` and vise versa. [fs2-reactive-streams](https://github.com/typelevel/fs2/tree/main/reactive-streams) is used for the conversion.

I only got rid of a conversion that converts from fs2 streams to Reactive Streams for `HttpResponse`.
Another remaining conversion is `HttpRequest`(Reactive Streams) to FS2 streams using fs2-reactive-streams.
I am going to remove `fs2-reactive-streams` dependency and implement an optimized version for fs2.

**Comment by ngbinh** [comment]
Thanks for the explanation. Looking forward to it!

**Comment by andyczerwonka** [comment]
> > Is it resolved?
> 
> I think there is a room we can optimize http4s-armeria performance. 💪
> 
> Loads of the benchmark were a simple "hello world" text message with HTTP/1.1. Because of the simple load, the bottleneck point which mostly consumes CPU was a conversion from Reactive Streams to fs2 `Stream` and vise versa. [fs2-reactive-streams](https://github.com/typelevel/fs2/tree/main/reactive-streams) is used for the conversion.
> 
> I only got rid of a conversion that converts from fs2 streams to Reactive Streams for `HttpResponse`. Another remaining conversion is `HttpRequest`(Reactive Streams) to FS2 streams using fs2-reactive-streams. I am going to remove `fs2-reactive-streams` dependency and implement an optimized version for fs2.

@ikhoon was the optimization to fs2 done?

**Comment by ikhoon** [comment]
It was partially done and is still in progress. But there were many performance optimizations on the Armeria side. 
I strongly believe that http4s-armeria can be used in production.

**Comment by danicheg** [comment]
@ikhoon It's been two years since you have done the benchmark. It'd be nice if you can run it with the new `http4s-armeria`/`blaze`. I believe some numbers must change in the benchmark results.

**Comment by ikhoon** [comment]
Agreed. Benchmark results might be changed somehow. I will run benchmarks soon.
